apps/emqx/src/emqx_persistent_message_ds_gc_worker.erl

```diff
@@ -0,0 +1,126 @@
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+-module(emqx_persistent_message_ds_gc_worker).
+
+-behaviour(gen_server).
+
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+-include_lib("stdlib/include/qlc.hrl").
+-include_lib("stdlib/include/ms_transform.hrl").
+
+-include("emqx_persistent_session_ds.hrl").
+
+%% API
+-export([
+    start_link/0
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+%% call/cast/info records
+-record(gc, {}).
+
+%%--------------------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------------------
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%%--------------------------------------------------------------------------------
+%% `gen_server' API
+%%--------------------------------------------------------------------------------
+
+init(_Opts) ->
+    ensure_gc_timer(),
+    State = #{},
+    {ok, State}.
+
+handle_call(_Call, _From, State) ->
+    {reply, error, State}.
+
+handle_cast(_Cast, State) ->
+    {noreply, State}.
+
+handle_info(#gc{}, State) ->
+    try_gc(),
+    ensure_gc_timer(),
+    {noreply, State};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+%%--------------------------------------------------------------------------------
+%% Internal fns
+%%--------------------------------------------------------------------------------
+
+ensure_gc_timer() ->
+    Timeout = emqx_config:get([session_persistence, message_retention_period]),
+    _ = erlang:send_after(Timeout, self(), #gc{}),
+    ok.
+
+try_gc() ->
+    %% Only cores should run GC.
+    CoreNodes = mria_membership:running_core_nodelist(),
+    Res = global:trans(
+        {?MODULE, self()},
+        fun() -> ?tp_span(debug, ds_message_gc, #{}, start_gc()) end,
+        CoreNodes,
+        %% Note: we set retries to 1 here because, in rare occasions, GC might start at the
+        %% same time in more than one node, and each one will abort the other.  By allowing
+        %% one retry, at least one node will (hopefully) get to enter the transaction and
+        %% the other will abort.  If GC runs too fast, both nodes might run in sequence.
+        %% But, in that case, GC is clearly not too costly, and that shouldn't be a problem,
+        %% resource-wise.
+        _Retries = 1
+    ),
+    case Res of
+        aborted ->
+            ?tp(ds_message_gc_lock_taken, #{}),
+            ok;
+        ok ->
+            ok
+    end.
+
+now_ms() ->
+    erlang:system_time(millisecond).
+
+start_gc() ->
+    ok = emqx_ds:add_generation(?PERSISTENT_MESSAGE_DB),
+    AllGens = emqx_ds:list_generations_with_lifetimes(?PERSISTENT_MESSAGE_DB),
```

Comment:
Not directly related to the GC, but adjacent: we want to have the ability to preserve the LTS between the generations. Let's brainstorm how it can be done.

Quick and dirty solution: the storage layer backend can query the metadata CF, and copy the data from its predecessor, if it has the same module. But maybe someone has a less hacky solution.

Preferably, it should be decided before the `list_generations` API is committed, since it _might_ require some passage of the data.

Reply:
It feels like this inheritance should be part of `add_generation`.  :thinking: 

Since the need for inheritance is storage layer dependent (i.e.: LTS only),  and since `add_generation` already simply creates a new gen with whatever is configured at the moment, it seems like this check for "if it's same module, inherit this" should be done there and not in the GC.

Reply:
Another dirty solution: we may keep LTS trie out of generation data and never GC it. Just apply thresholds from the current gen.

Reply:
That'll still require some coordination, because in that case the next generation should use the same column family to preserve the metadata.  Also, it might introduce a bit of complexity when the impl module changes: if we go from LTS to some other module, then the old trie column family should be dropped if we keep it around.

Reply:
> Another dirty solution: we may keep LTS trie out of generation data and never GC it. Just apply thresholds from the current gen when inserting and clean it up on demand.

This is an interesting idea, actually.